# Finding the perfect coffee neighborhood in Toronto 

### Part 1 : Retrieve the postal code, borough and beighborhood data from the Wiki 

Import the required libraries 

In [1]:
import pandas as pd
from bs4 import BeautifulSoup as bs
import requests as rq
import numpy as np 
import jsoｎ
from pandas.io.json import json_normalize
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
from geopy.geocoders import Nominatim
import folium
print('Libraries imported.')
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

Libraries imported.


Retreiving the URL and creating a Beautiful soup object

In [2]:
url ="http://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
response = rq.get(url)
soup = bs(response.text)

Retrieve the required intormation into a list 

In [3]:
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

Create a dataframe with the list we created 

In [4]:
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})
df.shape

(103, 3)

We get 103 rows of data in total 

### Part 2 : Retrieve the latitude and longitude for Toronto and each Borough

In [5]:
df_latlong = pd.read_csv(r'D:\Data.hhcheung\Desktop\Geospatial_Coordinates.csv') 

Joint the two dataframe

In [6]:
df_latlong.rename(columns ={'Postal Code':'PostalCode'}, inplace=True)
df_Merged = df.merge(df_latlong)
df_Merged.head()

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

In [7]:
address = 'Toronto, Toronto'

geolocator = Nominatim(user_agent="CAN_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.65238435, -79.38356765.


Display the map showing all the neighborhood 

In [8]:
# create map of Toronto using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_Merged['Latitude'], df_Merged['Longitude'], df_Merged['Borough'], df_Merged['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

### Part 3 :　Explore the neighborhoods in Toronto

Adding the foursquares API

In [9]:
CLIENT_ID = 'E2AOTYTD2K2WYLRVYREBCR0PV31KVPGRNWHPALCXBXZ4O3QU' # your Foursquare ID
CLIENT_SECRET = 'IBZRGPABQN20UIVU1QYX5K3IF35TYODHOCFXFTEESX1TNYSM' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: E2AOTYTD2K2WYLRVYREBCR0PV31KVPGRNWHPALCXBXZ4O3QU
CLIENT_SECRET:IBZRGPABQN20UIVU1QYX5K3IF35TYODHOCFXFTEESX1TNYSM


Function : Return the nearby venue for a neighborhood

In [10]:
def getNearbyVenues(names, latitudes, longitudes, radius=750):
    
    #Create a empty list 
    venues_list=[] 
    for name, lat, lng in zip(names, latitudes, longitudes):
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = rq.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Pass all the neighborhood into the function and create a new dataframe 

In [11]:
toronto_venues = getNearbyVenues(names=df_Merged['Neighborhood'],
                                   latitudes=df_Merged['Latitude'],
                                   longitudes=df_Merged['Longitude'])
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3         Parkwoods              43.753259              -79.329656   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude     Venue Category  
0         Brookbanks Park       43.751976       -79.332140               Park  
1           Variety Store       43.751974       -79.333114  Food & Drink Shop  
2       DVP at York Mills       43.758899       -79.334099       Intersection  
3         TTC Stop #09083       43.759655       -79.332223           Bus Stop  
4  Victoria Village Arena       43.723481       -79.315635       Hockey Arena

In [12]:
toronto_venues.shape

(3658, 7)

Count the number of unique categories among the venue returned

In [13]:
toronto_venues.groupby('Neighborhood').count()
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 324 uniques categories.


In [14]:
Venue_df = toronto_venues['Venue Category'].unique()
print(Venue_df)

['Park' 'Food & Drink Shop' 'Intersection' 'Bus Stop' 'Hockey Arena'
 'Portuguese Restaurant' 'Coffee Shop' 'Pizza Place' 'Playground'
 'Sporting Goods Shop' 'Bakery' 'Distribution Center' 'Historic Site'
 'Restaurant' 'Farmers Market' 'Chocolate Shop' 'Spa'
 'Mediterranean Restaurant' 'Breakfast Spot' 'Performing Arts Venue'
 'Gym / Fitness Center' 'Italian Restaurant' 'Pub' 'Liquor Store'
 'Dessert Shop' 'French Restaurant' 'Café' 'Tech Startup'
 'Thai Restaurant' 'Yoga Studio' 'Theater' 'Shoe Store' 'Animal Shelter'
 'Event Space' 'Pool' 'Gastropub' 'Art Gallery' 'Furniture / Home Store'
 'Brewery' 'Cosmetics Shop' 'Food Truck' 'Sandwich Place' 'Gym Pool' 'Bar'
 'Electronics Store' 'Sushi Restaurant' 'Bank' 'Beer Store' 'Skating Rink'
 'Pharmacy' 'Fast Food Restaurant' 'Hotel' 'Grocery Store'
 'Discount Store' 'Antique Shop' 'Tram Station' 'Athletics & Sports'
 'Dance Studio' 'Boutique' 'Vietnamese Restaurant' 'Greek Restaurant'
 'Fried Chicken Joint' 'Bowling Alley' 'Clothing Store

"Coffee Shop" & "Café" is the category we are looking for 

Show all the coffee shop retrieved 

In [155]:
Coffee_df = toronto_venues[toronto_venues['Venue Category'].isin(['Coffee Shop','Café'])]
Coffee_df

Neighborhood  \
6                                      Victoria Village   
13                            Regent Park, Harbourfront   
20                            Regent Park, Harbourfront   
22                            Regent Park, Harbourfront   
23                            Regent Park, Harbourfront   
24                            Regent Park, Harbourfront   
30                            Regent Park, Harbourfront   
38                            Regent Park, Harbourfront   
39                            Regent Park, Harbourfront   
43                            Regent Park, Harbourfront   
46                            Regent Park, Harbourfront   
56                            Regent Park, Harbourfront   
62                            Regent Park, Harbourfront   
77                            Regent Park, Harbourfront   
104                    Lawrence Manor, Lawrence Heights   
116                    Lawrence Manor, Lawrence Heights   
128                       Ontario Provincial Government   
135                       Ontario Provincial Government   
137                       Ontario Provincial Government   
141                       Ontario Provincial Government   
148                       Ontario Provincial Government   
167                       Ontario Provincial Government   
176                       Ontario Provincial Government   
179                       Ontario Provincial Government   
182                       Ontario Provincial Government   
186                       Ontario Provincial Government   
187                       Ontario Provincial Government   
189                       Ontario Provincial Government   
191                       Ontario Provincial Government   
193                       Ontario Provincial Government   
194                       Ontario Provincial Government   
196                       Ontario Provincial Government   
199                       Ontario Provincial Government   
204                                    Islington Avenue   
218                                      Malvern, Rouge   
221                                      Malvern, Rouge   
223                                     Don Mills North   
237                     Parkview Hill, Woodbine Gardens   
250                            Garden District, Ryerson   
260                            Garden District, Ryerson   
277                            Garden District, Ryerson   
288                            Garden District, Ryerson   
294                            Garden District, Ryerson   
297                            Garden District, Ryerson   
307                            Garden District, Ryerson   
313                            Garden District, Ryerson   
315                            Garden District, Ryerson   
320                            Garden District, Ryerson   
326                            Garden District, Ryerson   
329                            Garden District, Ryerson   
339                            Garden District, Ryerson   
352                                           Glencairn   
393                                     Don Mills South   
399                                     Don Mills South   
415                                      St. James Town   
425                                      St. James Town   
426                                      St. James Town   
437                                      St. James Town   
442                                      St. James Town   
445                                      St. James Town   
447                                      St. James Town   
459                                      St. James Town   
465                                      St. James Town   
472                                      St. James Town   
482                                      St. James Town   
492                                      St. James Town   
496                                      St. James Town   
500                            

In [156]:
Coffee_df.shape

(439, 7)

Toronto consists of 439 coffee shops ＆ Café. 

Display the map that indicate the coffee shop 

In [157]:
# create map of Toronto using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, venue, neighborhood in zip(Coffee_df['Venue Latitude'], Coffee_df['Venue Longitude'], Coffee_df['Venue'], Coffee_df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, venue)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

In [158]:
len(Coffee_df['Venue'].unique())

160

There are 160 unique coffee shop in total 

In [159]:
Coffee_Venue = Coffee_df.groupby('Venue').size().reset_index(name='counts')
Coffee_Venue.head()

Venue  counts
0          808 Social       2
1               9bars       2
2    AGO Espresso Bar       1
3       Absolute Cafe       1
4  Aroma Espresso Bar      10

In [160]:
Coffee_Chain = Coffee_Venue[Coffee_Venue.counts >5]
Coffee_Chain

Venue  counts
4       Aroma Espresso Bar      10
8          Balzac's Coffee      11
28             Café Plenty       7
44   Dineen @CommerceCourt       7
45           Dineen Coffee       7
76          Jimmy's Coffee       8
93                   Maman       6
103         Mos Mos Coffee       7
104         NEO COFFEE BAR       7
111  Pilot Coffee Roasters      12
125             Second Cup      15
130              Starbucks      68
150            Tim Hortons      71
153          Versus Coffee       6

In [205]:
Coffee_Chain_List = Coffee_Chain['Venue']
Coffee_Chain_List

4         Aroma Espresso Bar
8            Balzac's Coffee
28               Café Plenty
44     Dineen @CommerceCourt
45             Dineen Coffee
76            Jimmy's Coffee
93                     Maman
103           Mos Mos Coffee
104           NEO COFFEE BAR
111    Pilot Coffee Roasters
125               Second Cup
130                Starbucks
150              Tim Hortons
153            Versus Coffee
Name: Venue, dtype: object

In [161]:
Coffee_Chain.shape

(14, 2)

There are 14 Coffee Chain with shop number larger than 5 

Finding the neigborhood with the most coffee shop 

In [162]:
Coffee_Neigborhood = Coffee_df.groupby('Neighborhood').size().reset_index(name='counts').sort_values(by=['counts'],ascending=False)
Coffee_Neigborhood.head()

Neighborhood  counts
9                       Central Bay Street      19
51           Ontario Provincial Government      17
59                          St. James Town      17
28  First Canadian Place, Underground city      16
13          Commerce Court, Victoria Hotel      16

Central Bay Street are the neighborhood with the most coffee shop 

### Part 4 : Analysis the neighborhood 

It shows us how many venues there are in each Neighborhood:

In [164]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                              14   
Alderwood, Long Branch                                                 12   
Bathurst Manor, Wilson Heights, Downsview North                        25   
Bayview Village                                                        11   
Bedford Park, Lawrence Manor East                                      39   
Berczy Park                                                           100   
Birch Cliff, Cliffside West                                             7   
Brockton, Parkdale Village, Exhibition Place                           83   
CN Tower, King and Spadina, Railway Lands, Harb...                     26   
Caledonia-Fairbanks                                                    13   
Cedarbrae                                                              20   
Central Bay Street                                                    100   
Christie                                                               31   
Church and Wellesley                                                  100   
Clairville, Humberwood, Woodbine Downs, West Hu...                      3   
Clarks Corners, Tam O'Shanter, Sullivan                                21   
Cliffside, Cliffcrest, Scarborough Village West                         4   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             71   
Davisville North                                                       31   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        7   
Don Mills North                                                         6   
Don Mills South                                                        27   
Dorset Park, Wexford Heights, Scarborough Town ...                     13   
Downsview Central                                                       7   
Downsview East                                                          8   
Downsview Northwest                                                    21   
Downsview West                                                          8   
Dufferin, Dovercourt Village                                           33   
Enclave of L4W                                                         43   
Enclave of M4L                                                         57   
Enclave of M5E                                                        100   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                     11   
Fairview, Henry Farm, Oriole                                           38   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                                5   
Garden District, Ryerson                                              100   
Glencairn                                                              19   
Golden Mile, Clairlea, Oakridge                                        15   
Guildwood, Morningside, West Hill                                      10   
Harbourfront East, Union Station, Toronto Islands                      93   
High Park, The Junction South                                          60   
Hillcrest Village                                                      13   
Humber Summit                                                           5   
Humberlea, Emery                                                        4   
Humewood-Cedarvale                                                     14   
India Bazaar, The Beaches West                                         58   
Islington Avenue                                                       10   
Kennedy Park, Ionview, East Birchmount Park                            17   
Kensington Market, Chinatown, Grange Park                             100   
Kingsview Village, St. Phillips, Martin Grove G.

#### Get dummies for the categories of these venues in each neighborhood.

In [165]:
indexNames =  toronto_venues[toronto_venues['Venue Category'].isin(['Coffee Shop','Café'])].index
toronto_venues_exclude_CoffeeShop = pd.DataFrame(data = toronto_venues.drop(indexNames, inplace=False))
toronto_venues_exclude_CoffeeShop.shape

(3219, 7)

In [166]:
toronto_venues.shape

(3658, 7)

In [167]:
Toronto_onehot=pd.get_dummies(toronto_venues_exclude_CoffeeShop[['Venue Category']], prefix='', prefix_sep='')
Toronto_onehot['Neighborhood']=toronto_venues_exclude_CoffeeShop['Neighborhood']

fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0            0                  0               0                  0        0   
1            0                  0               0                  0        0   
2            0                  0               0                  0        0   
3            0                  0               0                  0        0   
4            0                  0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Amphitheater  Animal Shelter  \
0                 0                    0             0               0   
1                 0                    0             0               0   
2                 0                    0             0               0   
3                 0                    0             0               0   
4                 0                    0             0               0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Arts & Entertainment  Asian Restaurant  Athletics & Sports  \
0                     0                 0                   0   
1                     0                 0                   0   
2                     0                 0                   0   
3                     0                 0                   0   
4                     0                 0                   0   

   Auto Dealership  Auto Garage  Auto Workshop  Automotive Shop  BBQ Joint  \
0                0            0              0                0          0   
1                0            0              0                0          0   
2                0            0              0                0          0   
3                0            0              0                0          0   
4                0            0              0                0          0   

   Baby Store  Badminton Court  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0           0                0           0       0     0    0               0   
1           0                0           0       0     0    0               0   
2           0                0           0       0     0    0               0   
3           0                0           0       0     0    0               0   
4           0                0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  Beach Bar  \
0                 0                 0                   0      0          0   
1                 0                 0                   0      0          0   
2                 0                 0                   0      0          0   
3                 0                 0                   0      0          0   
4                 0                 0                   0      0          0   

   Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  Belgian Restaurant  \
0                0         0            0           0                   0   
1                0         0            0           0                   0   
2                0         0            0           0                   0   
3                0         0            0           0    

In [168]:
Toronto_onehot.shape

(3219, 322)

Analysis each neighborhood

In [169]:
toronto_grouped = Toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Yoga Studio  \
0                                        Agincourt          0.0   
1                           Alderwood, Long Branch          0.0   
2  Bathurst Manor, Wilson Heights, Downsview North          0.0   
3                                  Bayview Village          0.0   
4                Bedford Park, Lawrence Manor East          0.0   

   Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0                0.0             0.0                0.0      0.0   
1                0.0             0.0                0.0      0.0   
2                0.0             0.0                0.0      0.0   
3                0.0             0.0                0.0      0.0   
4                0.0             0.0                0.0      0.0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                 0.0           0.0             0.0              0.0   
1                 0.0           0.0             0.0              0.0   
2                 0.0           0.0             0.0              0.0   
3                 0.0           0.0             0.0              0.0   
4                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Amphitheater  Animal Shelter  \
0               0.0             0.000000           0.0             0.0   
1               0.0             0.000000           0.0             0.0   
2               0.0             0.000000           0.0             0.0   
3               0.0             0.000000           0.0             0.0   
4               0.0             0.028571           0.0             0.0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0           0.0       0.0          0.0         0.0                  0.0   
1           0.0       0.0          0.0         0.0                  0.0   
2           0.0       0.0          0.0         0.0                  0.0   
3           0.0       0.0          0.0         0.0                  0.0   
4           0.0       0.0          0.0         0.0                  0.0   

   Arts & Entertainment  Asian Restaurant  Athletics & Sports  \
0                   0.0               0.0                 0.0   
1                   0.0               0.0                 0.0   
2                   0.0               0.0                 0.0   
3                   0.0               0.0                 0.0   
4                   0.0               0.0                 0.0   

   Auto Dealership  Auto Garage  Auto Workshop  Automotive Shop  BBQ Joint  \
0              0.0          0.0            0.0              0.0        0.0   
1              0.0          0.0            0.0              0.0        0.0   
2              0.0          0.0            0.0              0.0        0.0   
3              0.0          0.0            0.0              0.0        0.0   
4              0.0          0.0            0.0              0.0        0.0   

   Baby Store  Badminton Court  Bagel Shop    Bakery      Bank  Bar  \
0         0.0         0.071429    0.000000  0.000000  0.000000  0.0   
1         0.0         0.000000    0.000000  0.000000  0.000000  0.0   
2         0.0         0.000000    0.000000  0.000000  0.086957  0.0   
3         0.0         0.000000    0.000000  0.000000  0.200000  0.0   
4         0.0         0.000000    0.028571  0.028571  0.028571  0.0   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0             0.0               0.0               0.0                 0.0   
1             0.0               0.0               0.0                 0.0   
2             0.0               0.0               0.0                 0.0   
3             0.0               0.0               0.0                 0.0   
4             0.0               0.0               0.0                 0.0   

   Beach  Beach Bar  Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  \
0    0.0        0.0              0.0       0.0          0.0         0.0 

Print the size of the new dataframe 

In [170]:
toronto_grouped.shape

(102, 322)

Create the new dataframe and display the top 10 venues for each neighborhood.

In [171]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [172]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood     1st Most Common Venue  \
0                                        Agincourt  Mediterranean Restaurant   
1                           Alderwood, Long Branch         Convenience Store   
2  Bathurst Manor, Wilson Heights, Downsview North               Pizza Place   
3                                  Bayview Village       Japanese Restaurant   
4                Bedford Park, Lawrence Manor East        Italian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Skating Rink    Seafood Restaurant       Badminton Court   
1           Pizza Place        Sandwich Place                   Pub   
2                  Bank        Ice Cream Shop     Mobile Phone Shop   
3                  Bank         Grocery Store          Intersection   
4            Restaurant        Sandwich Place     Indian Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0             Newsagent       Motorcycle Shop         Shopping Mall   
1                   Gym            Donut Shop           Gas Station   
2         Shopping Mall           Bridal Shop           Supermarket   
3         Shopping Mall          Skating Rink    Chinese Restaurant   
4               Butcher          Intersection            Bagel Shop   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0        Breakfast Spot      Sushi Restaurant                  Supermarket  
1          Skating Rink                  Park                Women's Store  
2      Sushi Restaurant                 Diner               Sandwich Place  
3            Playground   Dumpling Restaurant  Eastern European Restaurant  
4                Bakery                  Bank             Sushi Restaurant

## Part 5: Cluster the new dataframe 

In [173]:
# set number of clusters
kclusters = 7

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([6, 4, 4, 4, 6, 0, 6, 6, 6, 6])

In [174]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_Merged

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
toronto_merged # check the last columns!

PostalCode                 Borough  \
0          M3A              North York   
1          M4A              North York   
2          M5A        Downtown Toronto   
3          M6A              North York   
4          M7A            Queen's Park   
5          M9A               Etobicoke   
6          M1B             Scarborough   
7          M3B              North York   
8          M4B               East York   
9          M5B        Downtown Toronto   
10         M6B              North York   
11         M9B               Etobicoke   
12         M1C             Scarborough   
13         M3C              North York   
14         M4C               East York   
15         M5C        Downtown Toronto   
16         M6C                    York   
17         M9C               Etobicoke   
18         M1E             Scarborough   
19         M4E            East Toronto   
20         M5E        Downtown Toronto   
21         M6E                    York   
22         M1G             Scarborough   
23         M4G               East York   
24         M5G        Downtown Toronto   
25         M6G        Downtown Toronto   
26         M1H             Scarborough   
27         M2H              North York   
28         M3H              North York   
29         M4H               East York   
30         M5H        Downtown Toronto   
31         M6H            West Toronto   
32         M1J             Scarborough   
33         M2J              North York   
34         M3J              North York   
35         M4J  East York/East Toronto   
36         M5J        Downtown Toronto   
37         M6J            West Toronto   
38         M1K             Scarborough   
39         M2K              North York   
40         M3K              North York   
41         M4K            East Toronto   
42         M5K        Downtown Toronto   
43         M6K            West Toronto   
44         M1L             Scarborough   
45         M2L              North York   
46         M3L              North York   
47         M4L            East Toronto   
48         M5L        Downtown Toronto   
49         M6L              North York   
50         M9L              North York   
51         M1M             Scarborough   
52         M2M              North York   
53         M3M              North York   
54         M4M            East Toronto   
55         M5M              North York   
56         M6M                    York   
57         M9M              North York   
58         M1N             Scarborough   
59         M2N              North York   
60         M3N              North York   
61         M4N         Central Toronto   
62         M5N         Central Toronto   
63         M6N                    York   
64         M9N                    York   
65         M1P             Scarborough   
66         M2P              North York   
67         M4P         Central Toronto   
68         M5P         Central Toronto   
69         M6P            West Toronto   
70         M9P               Etobicoke   
71         M1R             Scarborough   
72         M2R              North York   
73         M4R         Central Toronto   
74         M5R         Central Toronto   
75         M6R            West Toronto   
76         M7R             Mississauga   
77         M9R               Etobicoke   
78         M1S             Scarborough   
79         M4S         Central Toronto   
80         M5S        Downtown Toronto   
81         M6S            West Toronto   
82         M1T             Scarborough   
83         M4T         Central Toronto   
84         M5T        Downtown Toronto   
85         M1V             Scarborough   
86         M4V         Central Toronto   
87         M5V        Downtown Toronto   
88         M8V               Etobicoke   
89         M9V               Etobicoke   
90         M1W             Scarborough   
91         M4W        Downtown Toronto   
92         M5W  Downtown Toronto Stn A   
93         M8W               Etobicoke   
94         M9W

In [175]:
indexNames2 =  toronto_merged[toronto_merged['Cluster Labels'].isna()].index
toronto_merged_Clean = pd.DataFrame(data =  toronto_merged.drop(indexNames2, inplace=False))
toronto_merged_Clean.head()

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  Cluster Labels   1st Most Common Venue 2nd Most Common Venue  \
0 -79.329656             5.0                Bus Stop          Intersection   
1 -79.315572             4.0            Intersection            Playground   
2 -79.360636             6.0              Restaurant                  Park   
3 -79.464763             6.0  Furniture / Home Store        Clothing Store   
4 -79.389494             6.0          Sandwich Place                  Park   

   3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0                   Park     Food & Drink Shop       Doner Restaurant   
1  Portuguese Restaurant           Pizza Place    Sporting Goods Shop   
2                 Bakery                   Pub                Theater   
3             Restaurant  Fast Food Restaurant  Vietnamese Restaurant   
4    Japanese Restaurant    Italian Restaurant       Sushi Restaurant   

  6th Most Common Venue  7th Most Common Venue 8th Most Common Venue  \
0    Dim Sum Restaurant                  Diner        Discount Store   
1          Hockey Arena                   Park               Dog Run   
2    Italian Restaurant  Performing Arts Venue        Breakfast Spot   
3          Dessert Shop       Sushi Restaurant    Athletics & Sports   
4           Yoga Studio          Burrito Place             Gastropub   

  9th Most Common Venue 10th Most Common Venue  
0   Distribution Center               Dive Bar  
1          Dessert Shop     Dim Sum Restaurant  
2       Thai Restaurant               Gym Pool  
3         Bowling Alley    Fried Chicken Joint  
4             Bookstore             Smoke Shop

In [176]:
toronto_merged_Clean['Cluster Labels']= toronto_merged_Clean['Cluster Labels'].astype(int)
toronto_merged_Clean.head()

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  Cluster Labels   1st Most Common Venue 2nd Most Common Venue  \
0 -79.329656               5                Bus Stop          Intersection   
1 -79.315572               4            Intersection            Playground   
2 -79.360636               6              Restaurant                  Park   
3 -79.464763               6  Furniture / Home Store        Clothing Store   
4 -79.389494               6          Sandwich Place                  Park   

   3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0                   Park     Food & Drink Shop       Doner Restaurant   
1  Portuguese Restaurant           Pizza Place    Sporting Goods Shop   
2                 Bakery                   Pub                Theater   
3             Restaurant  Fast Food Restaurant  Vietnamese Restaurant   
4    Japanese Restaurant    Italian Restaurant       Sushi Restaurant   

  6th Most Common Venue  7th Most Common Venue 8th Most Common Venue  \
0    Dim Sum Restaurant                  Diner        Discount Store   
1          Hockey Arena                   Park               Dog Run   
2    Italian Restaurant  Performing Arts Venue        Breakfast Spot   
3          Dessert Shop       Sushi Restaurant    Athletics & Sports   
4           Yoga Studio          Burrito Place             Gastropub   

  9th Most Common Venue 10th Most Common Venue  
0   Distribution Center               Dive Bar  
1          Dessert Shop     Dim Sum Restaurant  
2       Thai Restaurant               Gym Pool  
3         Bowling Alley    Fried Chicken Joint  
4             Bookstore             Smoke Shop

In [177]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_Clean['Latitude'], toronto_merged_Clean['Longitude'], toronto_merged_Clean['Neighborhood'], toronto_merged_Clean['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Cluster 1 

In [178]:
toronto_merged_Clean.loc[toronto_merged_Clean['Cluster Labels'] == 0, toronto_merged_Clean.columns[[1] + list(range(5, toronto_merged_Clean.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
7               North York               0   Japanese Restaurant   
9         Downtown Toronto               0                 Hotel   
15        Downtown Toronto               0        Cosmetics Shop   
20        Downtown Toronto               0                 Hotel   
24        Downtown Toronto               0        Clothing Store   
30        Downtown Toronto               0                 Hotel   
33              North York               0        Clothing Store   
36        Downtown Toronto               0                 Hotel   
42        Downtown Toronto               0                 Hotel   
48        Downtown Toronto               0                 Hotel   
92  Downtown Toronto Stn A               0                 Hotel   
97        Downtown Toronto               0                 Hotel   
99        Downtown Toronto               0   Japanese Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
7                    Gym  Caribbean Restaurant          Dessert Shop   
9              Gastropub   Japanese Restaurant        Clothing Store   
15    Seafood Restaurant                 Hotel             Gastropub   
20            Restaurant                  Park              Beer Bar   
24           Art Gallery   Japanese Restaurant      Ramen Restaurant   
30        Clothing Store               Theater              Beer Bar   
33   Japanese Restaurant            Restaurant                  Bank   
36         Boat or Ferry               Brewery   Japanese Restaurant   
42                   Gym            Restaurant    Seafood Restaurant   
48              Beer Bar      Asian Restaurant          Concert Hall   
92   Japanese Restaurant              Beer Bar            Restaurant   
97            Restaurant               Theater   Japanese Restaurant   
99      Sushi Restaurant            Restaurant               Gay Bar   

   5th Most Common Venue   6th Most Common Venue 7th Most Common Venue  \
7            Event Space                    Farm    Dim Sum Restaurant   
9       Sushi Restaurant                     Spa               Theater   
15   American Restaurant              Restaurant    Italian Restaurant   
20   Japanese Restaurant            Cocktail Bar                Bakery   
24           Pizza Place           Burrito Place                   Pub   
30            Restaurant  Furniture / Home Store                 Plaza   
33             Juice Bar    Fast Food Restaurant         Shopping Mall   
36                  Park                   Plaza        Scenic Lookout   
42   Japanese Restaurant     American Restaurant       Thai Restaurant   
48             Gastropub                     Gym            Restaurant   
92          Cocktail Bar                     Gym                Bakery   
97                   Gym                  Lounge        Cosmetics Shop   
99           Men's Store             Yoga Studio    Italian Restaurant   

     8th Most Common Venue    9th Most Common Venue     10th Most Common Venue  
7                    Diner           Discount Store        Distribution Center  
9   Furniture / Home Store  New American Restaurant         Falafel Restaurant  
15             Pizza Place   Furniture / Home Store                   Beer Bar  
20          Sandwich Place       Seafood Restaurant                Art Gallery  
24                     Bar        French Restaurant  Middle Eastern Restaurant  
30     Arts & Crafts Store         Sushi Restaurant                        Gym  
33              Food Court                   Bakery               Liquor Store  
36                     Bar               Restaurant              Deli / Bodega  
42                 Theater                   Lounge                      Plaza  
48          Cosmetics Shop       Seafood Restaurant        Japanese Restaurant  
92          Farmers Market              Art Gallery                Cheese Shop  
97         Thai Restaurant       Seafood Restaur

#### Cluster 2

In [179]:
toronto_merged_Clean.loc[toronto_merged_Clean['Cluster Labels'] == 1, toronto_merged_Clean.columns[[1] + list(range(5, toronto_merged_Clean.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
62  Central Toronto               1            Playground   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
62                  Pool                Garden      Doner Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
62          Dessert Shop    Dim Sum Restaurant                 Diner   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
62        Discount Store   Distribution Center               Dive Bar

#### Cluster 3

In [180]:
toronto_merged_Clean.loc[toronto_merged_Clean['Cluster Labels'] == 2, toronto_merged_Clean.columns[[1] + list(range(5, toronto_merged_Clean.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
45  North York               2                  Pool   Martial Arts School   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
45            Donut Shop    Dim Sum Restaurant                 Diner   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
45        Discount Store   Distribution Center              Dive Bar   

   9th Most Common Venue 10th Most Common Venue  
45               Dog Run       Doner Restaurant

#### Cluster 4

In [181]:
toronto_merged_Clean.loc[toronto_merged_Clean['Cluster Labels'] == 3, toronto_merged_Clean.columns[[1] + list(range(5, toronto_merged_Clean.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
94  Etobicoke Northwest               3                 Hotel   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
94   Rental Car Location                Lounge   Dumpling Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
94                 Diner        Discount Store   Distribution Center   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
94              Dive Bar               Dog Run       Doner Restaurant

#### Cluster 5

In [182]:
toronto_merged_Clean.loc[toronto_merged_Clean['Cluster Labels'] == 4, toronto_merged_Clean.columns[[1] + list(range(5, toronto_merged_Clean.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1               North York               4          Intersection   
5                Etobicoke               4              Pharmacy   
6              Scarborough               4  Fast Food Restaurant   
8                East York               4           Pizza Place   
10              North York               4           Pizza Place   
11               Etobicoke               4           Pizza Place   
17               Etobicoke               4                  Park   
18             Scarborough               4  Fast Food Restaurant   
27              North York               4        Sandwich Place   
28              North York               4           Pizza Place   
32             Scarborough               4        Ice Cream Shop   
34              North York               4   Japanese Restaurant   
35  East York/East Toronto               4           Pizza Place   
38             Scarborough               4           Bus Station   
39              North York               4   Japanese Restaurant   
46              North York               4                  Park   
59              North York               4     Korean Restaurant   
60              North York               4  Fast Food Restaurant   
63                    York               4           Pizza Place   
64                    York               4                 Diner   
67         Central Toronto               4           Pizza Place   
70               Etobicoke               4           Flea Market   
71             Scarborough               4           Pizza Place   
72              North York               4           Pizza Place   
77               Etobicoke               4           Pizza Place   
82             Scarborough               4              Pharmacy   
85             Scarborough               4           Pizza Place   
88               Etobicoke               4              Pharmacy   
89               Etobicoke               4         Grocery Store   
90             Scarborough               4           Pizza Place   
93               Etobicoke               4     Convenience Store   

        2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
1                  Playground  Portuguese Restaurant           Pizza Place   
5                  Playground           Home Service         Grocery Store   
6                Home Service                    Spa                 Trail   
8                Home Service              Gastropub           Flea Market   
10              Grocery Store            Gas Station            Restaurant   
11                    Theater     Mexican Restaurant                   Gym   
17                 Beer Store            Pizza Place              Pharmacy   
18                Pizza Place       Greek Restaurant            Sports Bar   
27                 Restaurant           Tennis Court              Pharmacy   
28                       Bank         Ice Cream Shop     Mobile Phone Shop   
32                 Playground            Pizza Place  Fast Food Restaurant   
34         Miscellaneous Shop            Pizza Place  Fast Food Restaurant   
35       Fast Food Restaurant              Pet Store      Greek Restaurant   
38         Light Rail Station           Intersection   Rental Car Location   
39                       Bank          Grocery Store          Intersection   
46                Pizza Place  Vietnamese Restaurant         Shopping Mall   
59                Pizza Place       Ramen Restaurant      Sushi Restaurant   
60             Discount Store               Pharmacy         Grocery Store   
63                    Brewery     Athletics & Sports                  Park   
64        Fried Chicken Joint            Pizza Place              Pharmacy   
67                        Gym                   Park              Pharmacy   
70                 Playground            Pizza Place        Sandwich Place   
71  Middle Eastern Restaurant           Intersec

#### Clsuter 6

In [183]:
toronto_merged_Clean.loc[toronto_merged_Clean['Cluster Labels'] == 5, toronto_merged_Clean.columns[[1] + list(range(5, toronto_merged_Clean.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0          North York               5              Bus Stop   
14          East York               5                  Park   
22        Scarborough               5                  Park   
52         North York               5                  Park   
56               York               5           Gas Station   
57         North York               5           Gas Station   
66         North York               5                  Park   
68    Central Toronto               5         Jewelry Store   
91   Downtown Toronto               5                  Park   
101         Etobicoke               5            Playground   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Intersection                  Park     Food & Drink Shop   
14     Athletics & Sports          Skating Rink              Bus Line   
22      Convenience Store      Business Service            Donut Shop   
52            Pizza Place      Sushi Restaurant                  Bank   
56      Convenience Store          Intersection        Sandwich Place   
57         Baseball Field        Discount Store     Convenience Store   
66          Bowling Alley     Convenience Store                   Gym   
68                   Park  Gym / Fitness Center                 Trail   
91                  Trail            Playground           Candy Store   
101  Gym / Fitness Center        Baseball Field                  Park   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Doner Restaurant    Dim Sum Restaurant                 Diner   
14               Bus Stop          Intersection           Curling Ice   
22                  Diner        Discount Store   Distribution Center   
52         Sandwich Place     Indian Restaurant                 Trail   
56         Discount Store                  Park          Dessert Shop   
57     Falafel Restaurant            Donut Shop                 Diner   
66              Pet Store                 Diner        Discount Store   
68       Sushi Restaurant          Dessert Shop    Dim Sum Restaurant   
91          Women's Store    Dim Sum Restaurant                 Diner   
101         Women's Store      Doner Restaurant    Dim Sum Restaurant   

           8th Most Common Venue 9th Most Common Venue  \
0                 Discount Store   Distribution Center   
14                    Beer Store     Convenience Store   
22                      Dive Bar               Dog Run   
52           Sporting Goods Shop             Pet Store   
56   Eastern European Restaurant   Dumpling Restaurant   
57                Farmers Market   Distribution Center   
66           Distribution Center              Dive Bar   
68                         Diner        Discount Store   
91                Discount Store   Distribution Center   
101                        Diner        Discount Store   

          10th Most Common Venue  
0                       Dive Bar  
14   Eastern European Restaurant  
22              Doner Restaurant  
52                    Comic Shop  
56                    Donut Shop  
57                      Dive Bar  
66                       Dog Run  
68           Distribution Center  
91                      Dive Bar  
101          Distribution Center

#### Cluster 7 

In [184]:
toronto_merged_Clean.loc[toronto_merged_Clean['Cluster Labels'] == 6, toronto_merged_Clean.columns[[1] + list(range(5, toronto_merged_Clean.shape[1]))]]

Borough  Cluster Labels     1st Most Common Venue  \
2         Downtown Toronto               6                Restaurant   
3               North York               6    Furniture / Home Store   
4             Queen's Park               6            Sandwich Place   
12             Scarborough               6            Breakfast Spot   
13              North York               6                       Gym   
16                    York               6                     Field   
19            East Toronto               6                       Pub   
21                    York               6                      Park   
23               East York               6    Furniture / Home Store   
25        Downtown Toronto               6             Grocery Store   
26             Scarborough               6         Indian Restaurant   
29               East York               6         Indian Restaurant   
31            West Toronto               6                  Pharmacy   
37            West Toronto               6                       Bar   
40              North York               6                      Park   
41            East Toronto               6          Greek Restaurant   
43            West Toronto               6                Restaurant   
44             Scarborough               6              Intersection   
47            East Toronto               6         Indian Restaurant   
49              North York               6   Health & Beauty Service   
50              North York               6               IT Services   
51             Scarborough               6               Wings Joint   
53              North York               6            Baseball Field   
54            East Toronto               6                       Bar   
55              North York               6        Italian Restaurant   
58             Scarborough               6                      Park   
61         Central Toronto               6          Business Service   
65             Scarborough               6         Indian Restaurant   
69            West Toronto               6           Thai Restaurant   
73         Central Toronto               6       Sporting Goods Shop   
74         Central Toronto               6                       Pub   
75            West Toronto               6                       Bar   
76             Mississauga               6        Chinese Restaurant   
78             Scarborough               6  Mediterranean Restaurant   
79         Central Toronto               6        Italian Restaurant   
80        Downtown Toronto               6                Restaurant   
81            West Toronto               6                    Bakery   
83         Central Toronto               6                      Park   
84        Downtown Toronto               6                       Bar   
86         Central Toronto               6          Sushi Restaurant   
87        Downtown Toronto               6       Rental Car Location   
96        Downtown Toronto               6               Pizza Place   
98               Etobicoke               6            Breakfast Spot   
100  East Toronto Business               6      Fast Food Restaurant   
102              Etobicoke               6      Gym / Fitness Center   

             2nd Most Common Venue          3rd Most Common Venue  \
2                             Park                         Bakery   
3                   Clothing Store                     Restaurant   
4                             Park            Japanese Restaurant   
12                             Bar             Italian Restaurant   
13                      Restaurant                   Intersection   
16                            Park               Sushi Restaurant   
19                  Sandwich Place            Japanese Restaurant   
21                   Women's Store                         Hostel   
23               Electronics Store            Sporting Goods Shop   
25              

In [185]:
toronto_merged_Clean.loc[toronto_merged_Clean['Neighborhood'] =='Central Bay Street']

PostalCode           Borough        Neighborhood   Latitude  Longitude  \
24        M5G  Downtown Toronto  Central Bay Street  43.657952 -79.387383   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
24               0        Clothing Store           Art Gallery   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
24   Japanese Restaurant      Ramen Restaurant           Pizza Place   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
24         Burrito Place                   Pub                   Bar   

   9th Most Common Venue     10th Most Common Venue  
24     French Restaurant  Middle Eastern Restaurant

Central Bay Street belongs to Cluster 1. 

#### Neighborhood that are in the same cluster group : 

In [197]:
Targeted_neighborhood =  toronto_merged_Clean.loc[toronto_merged_Clean['Cluster Labels'] == 0]['Neighborhood']
Targeted_neighborhood

7                                       Don Mills North
9                              Garden District, Ryerson
15                                       St. James Town
20                                          Berczy Park
24                                   Central Bay Street
30                             Richmond, Adelaide, King
33                         Fairview, Henry Farm, Oriole
36    Harbourfront East, Union Station, Toronto Islands
42             Toronto Dominion Centre, Design Exchange
48                       Commerce Court, Victoria Hotel
92                                       Enclave of M5E
97               First Canadian Place, Underground city
99                                 Church and Wellesley
Name: Neighborhood, dtype: object

In [204]:
Coffee_targeted_neighborhood = Coffee_Neigborhood[Coffee_Neigborhood['Neighborhood'].isin(Targeted_neighborhood)]
Coffee_targeted_neighborhood 

Neighborhood  counts
9                                  Central Bay Street      19
59                                     St. James Town      17
28             First Canadian Place, Underground city      16
13                     Commerce Court, Victoria Hotel      16
70           Toronto Dominion Centre, Design Exchange      15
25                                     Enclave of M5E      15
55                           Richmond, Adelaide, King      14
29                           Garden District, Ryerson      13
32  Harbourfront East, Union Station, Toronto Islands      12
4                                         Berczy Park      12
11                               Church and Wellesley      11
27                       Fairview, Henry Farm, Oriole       5
16                                    Don Mills North       1

Don Mills North have the least number of coffee shop, it's recommended to start a coffee shop in that neighborhood 

#### Neighborhood with healthy competition (With least chain store) 

In [208]:
#Drop all chain store 
indexNames3 =  Coffee_df[Coffee_df['Venue'].isin(Coffee_Chain_List)].index
Coffee_df_without_chain = pd.DataFrame(data =   Coffee_df.drop(indexNames3, inplace=False))
#Group by Neighborhood
Coffee_Neigborhood_without_chain = Coffee_df_without_chain.groupby('Neighborhood').size().reset_index(name='counts').sort_values(by=['counts'],ascending=False)
Coffee_Neigborhood_without_chain.head()
#Filter out the cluster group 1 
Coffee_targeted_neighborhood_without_chain = Coffee_Neigborhood_without_chain[Coffee_Neigborhood_without_chain['Neighborhood'].isin(Targeted_neighborhood)]
Coffee_targeted_neighborhood_without_chain

Neighborhood  counts
6                                  Central Bay Street       9
36                           Richmond, Adelaide, King       8
18             First Canadian Place, Underground city       6
38                                     St. James Town       6
19                           Garden District, Ryerson       5
16                                     Enclave of M5E       4
20  Harbourfront East, Union Station, Toronto Islands       4
9                      Commerce Court, Victoria Hotel       4
8                                Church and Wellesley       4
46           Toronto Dominion Centre, Design Exchange       4
1                                         Berczy Park       3
12                                    Don Mills North       1

Richmond, Adelaide, King have the healthiest competition environment.